File: programs/circular_queue.c
#include <stdio.h>

//variables.
int maxsize, front = -1, rear = -1, Q[50], choice, exited = 0, item;

//operations.
void insert();
void delete();
void display();

void makeSelection();

void main()
{
    //get maxszie of Queue.
    printf("Enter the size of Queue : ");
    scanf("%d", &maxsize);

    //do while loop for executing multiple operations on a single execution.
    do {

        // show available operations.
        printf("\n\n --- Available Operations --- \n\n");
        printf("1). Insertion \n");
        printf("2). Deletion \n");
        printf("3). Display \n");
        printf("4). Exit \n");

        printf("Enter your choice : \n");
        scanf("%d", &choice);

        makeSelection();

    } while(exited != 1);
}


void makeSelection()
{
    switch(choice){
        case 1 :
            insert();
        break;
        case 2 :
            delete();
        break;
        case 3 :
            display();
        break;
        default :
            printf("Invalid selection, Please try again...\n");
    }
}


void insert()
{
    if(front == (rear+1) % maxsize)
    {
        printf("\n\n Queue is full. \n\n");
    } else {
        if(front == -1 && rear == -1)
        {
            front++;
        }

        //read item.
        rear  = (rear+1) % maxsize;
        printf("\n Enter the item : ");
        scanf("%d", &item);
        Q[rear] = item;
        printf("%d inserted to the Queue", item);
    }
}

void delete()
{
    if(front == -1)
    {
        printf("\n\n Queue is empty \n\n");
    } else {
        if(front == rear)
        {
            printf("\n\n Queue resetted.\n\n");
        }

        item = Q[front];
        printf("%d is deleted from Queue");
        front = (front + 1) % 4;
    }
}

void display()
{
    
}

File: programs/CircularQueue.java
// Circular Queue implementation in Java.

import java.util.Scanner;

public class CircularQueue {
    int front, rear, size;
    int[] queue;

    public CircularQueue(int size) {
        this.size = size;
        front = -1;
        rear = -1;
        queue = new int[size];
    }

    public void insert(int data) {
        if (front == -1 && rear == -1) {
            front = rear = 0;
            queue[rear] = data;
        } else if ((rear + 1) % size == front) {
            System.out.println("Queue is full...");
        } else {
            rear = (rear + 1) % size;
            queue[rear] = data;
        }
    }

    public void delete() {
        if (front == -1 && rear == -1) {
            System.out.println("Queue is empty...");
        } else if (front == rear) {
            front = rear = -1;
        } else {
            front = (front + 1) % size;
        }
    }

    public void display() {
        if (front == -1 && rear == -1) {
            System.out.println("Queue is empty..");
        } else {
            int i = front;
            while (i != rear) {
                System.out.print(queue[i] + " ");
                i = (i + 1) % size;
            }
            System.out.println(queue[rear]);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the size of the queue");
        int size = sc.nextInt();
        CircularQueue cq = new CircularQueue(size);
        int choice = 0;
        while (choice != 4) {
            System.out.println("1. Insert");
            System.out.println("2. Delete");
            System.out.println("3. Display");
            System.out.println("4. Exit");
            System.out.println("Enter your choice : ");
            choice = sc.nextInt();
            switch (choice) {
                case 1:
                    System.out.println("Enter the data to be inserted : ");
                    int data = sc.nextInt();
                    cq.insert(data);
                    break;
                case 2:
                    cq.delete();
                    break;
                case 3:
                    cq.display();
                    break;
                case 4:
                    System.out.println("Exiting....");
                    break;
                default:
                    System.out.println("Invalid choice...");
            }
        }
    }
}